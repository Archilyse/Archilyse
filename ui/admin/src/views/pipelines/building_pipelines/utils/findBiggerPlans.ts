import { C } from 'Common';
import { ProviderRequest } from 'Providers';

// @TODO: All of this to an util file
type PlanSize = {
  id: string;
  image_width: number;
  image_height: number;
};

const fetchPlanSizes = async (pipelines, masterPlan): Promise<PlanSize[]> => {
  const requests = pipelines
    .filter(pipeline => Number(pipeline.building_id) === Number(masterPlan.building_id)) // Only ones from the same building
    .map(pipelines => ProviderRequest.get(C.ENDPOINTS.PLAN(pipelines.id)));
  const plans = await Promise.all(requests);
  return plans.map((p: any) => ({ image_width: p.image_width, image_height: p.image_height, id: p.id }));
};

const findBiggerPlans = async (pipelines, masterPlan: { id: number; building_id: number }): Promise<string[]> => {
  const planSizes = await fetchPlanSizes(pipelines, masterPlan);
  const currentSize = planSizes.find(plan => String(plan.id) === String(masterPlan.id));

  const biggerPlans = planSizes.filter(
    plan => currentSize.image_width < plan.image_width || currentSize.image_height < plan.image_height
  );

  return biggerPlans.map(p => p.id);
};

export default findBiggerPlans;
