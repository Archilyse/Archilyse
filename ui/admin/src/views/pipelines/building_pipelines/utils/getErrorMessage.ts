import { displayValidationHttpError, isValidationError } from 'Components/ErrorBoundary';

const getErrorMessage = error => {
  if (isValidationError(error)) {
    return displayValidationHttpError(error);
  }
  return (error.response && error.response.data && error.response.data.msg) || error?.toString() || '';
};

export default getErrorMessage;
