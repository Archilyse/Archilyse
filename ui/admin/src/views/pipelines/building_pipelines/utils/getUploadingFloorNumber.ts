import UploadingFloor from '../types/UploadingFloor';

const getUploadingFloorNumber = (data: UploadingFloor['data']) =>
  data.floor_upper_range !== data.floor_lower_range
    ? `${data.floor_lower_range} - ${data.floor_upper_range}`
    : data.floor_lower_range;

export default getUploadingFloorNumber;
