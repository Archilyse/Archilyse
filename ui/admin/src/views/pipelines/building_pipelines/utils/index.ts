export { default as getErrorMessage } from './getErrorMessage';
export { default as findBiggerPlans } from './findBiggerPlans';
export { default as getUploadingFloorNumber } from './getUploadingFloorNumber';
