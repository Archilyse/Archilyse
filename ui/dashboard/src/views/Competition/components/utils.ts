export const getSafePercents = (percents: number): [string, number] => {
  if (typeof percents !== 'number') return ['â€”', 0];

  const roundedPercents = Math.round(percents * 1000) / 10; // 0.xxx to xx.x
  if (roundedPercents > 100) {
    return ['100%', 100];
  }
  if (roundedPercents < 0) {
    return ['0%', 0];
  }

  return [`${roundedPercents}%`, roundedPercents];
};

export function checkIsExpanded(categoryKey: string, expandedCategories: string[]): boolean {
  return expandedCategories.some(item => {
    // key and path is a string like 'xxx.yyy.zzz'
    const categoryPath = categoryKey.split('.');
    const expandedPath = item.split('.');

    return categoryPath.every((_, index) => categoryPath[index] === expandedPath[index]);
  });
}
