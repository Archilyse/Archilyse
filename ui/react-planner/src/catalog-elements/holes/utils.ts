type DoorSweepingPoints = {
  angle_point: [number, number];
  closed_point: [number, number];
  opened_point: [number, number];
};

export function getArcSvgPathFromSweepingPoints(doorPoints: DoorSweepingPoints, holeLength: number) {
  /**
   * Parametrized the SVG path string as in http://xahlee.info/js/svg_path_ellipse_arc.html
   */
  const ellipseRadius = `${holeLength},${holeLength}`;
  const rotation = 0;
  const largeArcFlag = 0;
  const sweepFlag = getSweepFlag(doorPoints);
  const finishingPoint = `${doorPoints.opened_point}`;
  return `M${doorPoints.closed_point}  A${ellipseRadius} ${rotation} ${largeArcFlag},${sweepFlag} ${finishingPoint}`;
}

function getSweepFlag(doorPoints: DoorSweepingPoints) {
  /**
   * Specifies on which side of the path (from closed point -> opened point) the curve is drawn. Depends on where the angle (axis) point is.
   * A positive cross product between v1 & v2 corresponds to angle point on the left side of the path.
   */
  const v1 = [
    doorPoints.opened_point[0] - doorPoints.closed_point[0],
    doorPoints.opened_point[1] - doorPoints.closed_point[1],
  ]; //vector connecting closed point with opened point
  const v2 = [
    doorPoints.angle_point[0] - doorPoints.closed_point[0],
    doorPoints.angle_point[1] - doorPoints.closed_point[1],
  ]; //vector connecting closed point with angle point
  const cross_product = v1[0] * v2[1] - v1[1] * v2[0];
  if (cross_product > 0) {
    return 1;
  } else {
    return 0;
  }
}
