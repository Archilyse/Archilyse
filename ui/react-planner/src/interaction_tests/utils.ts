import thunk from 'redux-thunk';
// Mock Annotations response contains almost the same data as in tests/fixtures/react_planner_annotations/plan_background_image.json
// that is loaded in db_fixtures_ui
import { fireEvent, screen, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { applyMiddleware, compose, createStore } from 'redux';
import configureStore from 'redux-mock-store';
import {
  MOCK_ANNOTATIONS_RESPONSE,
  MOCK_AUTHENTICATION,
  MOCK_PLAN_RESPONSE,
  MOCK_SITE_STRUCTURE,
  serverMocks,
} from '../tests/utils/';
import { Models as PlannerModels, Plugins as PlannerPlugins, reducer as PlannerReducer } from '../export'; //react-planner
import MyCatalog from '../catalog-elements/mycatalog';
import Item from '../class/item';
import cloneDeep from '../utils/clone-deep';

const { buildHandler, ENDPOINTS_PATTERN } = serverMocks;

export const VIEWER_REGEXP = /viewer-[0-9]+x[0-9]+/;
export const EXPECTED_SCENE_WIDTH = 2187;
export const EXPECTED_SCENE_HEIGTH = 1640;

export const selectFromCatalog = async (container, itemName) => {
  let catalogToolbar = screen.queryByTestId('catalog-toolbar');
  if (!catalogToolbar) {
    await waitFor(() => expect(container.querySelector(`[tooltip="Open catalog"]`)).toBeTruthy(), {
      timeout: 3000,
    });
    userEvent.click(container.querySelector(`[tooltip="Open catalog"]`));
    catalogToolbar = screen.queryByTestId('catalog-toolbar');
  }
  const [catalogElement] = await within(catalogToolbar).findAllByText(new RegExp(itemName));
  userEvent.click(catalogElement);
};

export const clicksOnAnArea = () => {
  const [firstArea] = screen.getAllByTestId(/viewer-area-/);
  userEvent.click(firstArea);
};

export const hoverAnArea = () => {
  const [firstArea] = screen.getAllByTestId(/viewer-area-/);
  userEvent.hover(firstArea);
};

export const clickAndDragAMainVertex = store => {
  const allLines = Object.values(store.getState()['react-planner'].scene.layers['layer-1'].lines) as any;
  const firstLine = allLines[0];
  const lineId = firstLine.id;
  const mainVertexId = firstLine.vertices[0];

  const lineIdElem = screen.getByTestId(`viewer-line-${lineId}`);
  userEvent.click(lineIdElem);

  const mainVertexElem = screen.getByTestId(`vertex-${mainVertexId}`);
  fireEvent.mouseDown(mainVertexElem);
  fireEvent.mouseMove(mainVertexElem);
  fireEvent.mouseUp(mainVertexElem);
};

export const getLineBySeparatorType = (separatorType): any => {
  const layer = MOCK_ANNOTATIONS_RESPONSE.data.layers['layer-1'];
  const line = Object.values(layer.lines).find(line => line.type === separatorType);
  return line;
};

export const MockItemPositionValid = isValid => {
  const itemSpy = jest.spyOn(Item, 'isSelectedItemInValidPosition');
  itemSpy.mockImplementation(() => {
    return isValid;
  });
};

export const expectSnackbarWithMessage = async message => {
  await waitFor(() => expect(screen.getByText(new RegExp(message, 'i'))).toBeInTheDocument(), {
    timeout: 3000,
  });
};

export const assertBackgroundDimensions = async (width, height) => {
  const floorplanImg = screen.getByTestId('floorplan-img');
  expect(floorplanImg).toBeInTheDocument();
  await waitFor(() => expect(floorplanImg).toHaveAttribute('width', String(width)), { timeout: 6000 });
  await waitFor(() => expect(floorplanImg).toHaveAttribute('height', String(height)), { timeout: 6000 });
};

export const assertBackgroundTransform = async expectedTransform => {
  const backgroundImage = screen.getByTestId('background-img-group');
  await waitFor(() => expect(backgroundImage).toHaveAttribute('transform', expectedTransform), { timeout: 6000 });
};

export const assertBackgroundPosition = async (expectedX: number, expectedY: number) => {
  const backgroundImage = screen.getByTestId('floorplan-img');
  await waitFor(() => expect(backgroundImage).toHaveAttribute('x', String(expectedX)), { timeout: 6000 });
  await waitFor(() => expect(backgroundImage).toHaveAttribute('y', String(expectedY)), { timeout: 6000 });
};

export const assertSceneDimensions = async (width, height) => {
  await waitFor(() => expect(screen.getByTestId(new RegExp(`viewer-${width}x${height}`))).toBeInTheDocument(), {
    timeout: 3000,
  });
};

export const waitForSceneLoad = async () => {
  await waitFor(() => expect(screen.getByTestId(VIEWER_REGEXP)).toBeInTheDocument(), { timeout: 6000 });
  const viewer = screen.getByTestId(VIEWER_REGEXP);
  await waitFor(() => expect(viewer).toHaveAttribute('class', 'floorplan-img-loaded centered'), { timeout: 6000 });
  await assertBackgroundDimensions(EXPECTED_SCENE_WIDTH, EXPECTED_SCENE_HEIGTH);
  await waitFor(() => expect(screen.queryAllByTestId(/viewer-line-/).length).toBeTruthy(), { timeout: 6000 });
};

export const setupServer = server => {
  server.listen();
  server.use(
    ...[
      buildHandler(ENDPOINTS_PATTERN.FLOORPLAN_IMG_PLAN, 'get', new Blob(), 200),
      buildHandler(ENDPOINTS_PATTERN.ANNOTATION_PLAN, 'get', MOCK_ANNOTATIONS_RESPONSE, 200),
      buildHandler(ENDPOINTS_PATTERN.PLAN_BY_ID, 'get', MOCK_PLAN_RESPONSE, 200),
      buildHandler(ENDPOINTS_PATTERN.SITE_STRUCTURE, 'get', MOCK_SITE_STRUCTURE),
    ]
  );
};

// @TODO: Rename to `setupHelpers` mocks (jest.mock()) cannot be placed here as inside of the body of a function they won't work.
export const setupMocks = () => {
  // helper to find the new sizes => console.log(screen.getByText((content, element) => content.endsWith('mÂ²')));
  window.URL.createObjectURL = jest.fn(() => 'blob:http://localhost:9000/2b9be933-9b4f-4ec5-9d16-337423936ff9');
};

export const setupRedux = (withActionsHistory = false) => {
  MOCK_AUTHENTICATION();

  // Real state & reducer used in renderer.jsx
  const AppState = {
    'react-planner': new PlannerModels.State(),
  };

  const reducer = (state, action) => {
    state = state || cloneDeep(AppState);
    state = PlannerReducer(state['react-planner'], action);
    state = {
      'react-planner': state,
    };
    return state;
  };

  const plugins = [PlannerPlugins.Keyboard()];

  const props = {
    catalog: MyCatalog,
    width: 3000,
    height: 2000,
    plugins,
    stateExtractor: state => state['react-planner'],
  };

  const mockStore = configureStore([thunk]);
  const mockedStore = mockStore(AppState);

  const originalStore = createStore(reducer, null, compose(applyMiddleware(thunk)));
  let store = originalStore;

  // extend current store with other functions from 'redux-mock-store'
  // like `getActions()`
  if (withActionsHistory) {
    store = {
      ...mockedStore,
      ...originalStore,
    };
    const originalDispatch = originalStore.dispatch;
    store.dispatch = function (action) {
      mockedStore.dispatch(action);
      return originalDispatch(action);
    };
  }
  return { props, store };
};
