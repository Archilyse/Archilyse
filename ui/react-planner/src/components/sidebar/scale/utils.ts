import {
  PAGE_SIZES,
  SCALING_BASELINE,
  SCALING_MAX_DIFF_BETWEEN_MEASUREMENTS,
  SeparatorsType,
} from '../../../constants';
import ScaleMeasurement from '../../../types/ScaleMeasurement';
import {
  convertPixelsToCMs,
  getAreaSizeInPixels,
  getLineDistance,
  isLine,
  isPolygon,
  orderVertices,
} from '../../../utils/geometry';
import { getSelectedLayer } from '../../../utils/state-utils';

const isMeasurementOutlier = (measurements: ScaleMeasurement[], currentMeasurement) => {
  if (!measurements || !currentMeasurement) return false;

  const currentScaleFactor = calculateScale(currentMeasurement);
  const currentMeasurementDistanceInCm = convertPixelsToCMs(currentScaleFactor, SCALING_BASELINE);
  const differsTooMuchFromOtherMeasurements = measurements.some(m => {
    const scaleFactor = calculateScale(m);
    const distInCM = convertPixelsToCMs(scaleFactor, SCALING_BASELINE);
    return Math.abs(distInCM - currentMeasurementDistanceInCm) > SCALING_MAX_DIFF_BETWEEN_MEASUREMENTS;
  });
  return differsTooMuchFromOtherMeasurements;
};

const needsRepeating = measurements => {
  const [lastMeasurement] = measurements.slice(-1);
  return isMeasurementOutlier(measurements, lastMeasurement);
};

// The scale calculated is always the area scale factor, hence the Math.pow in the line section
const calculateScale = ({ points, distance, areaSize, area }: ScaleMeasurement): number => {
  if (isLine(points)) {
    const currentDistance = getLineDistance(points);
    return !isNaN(currentDistance) && currentDistance ? Math.pow(distance / currentDistance, 2) : 0;
  } else if (isPolygon(points) && area) {
    const currentAreaSize = getAreaSizeInPixels(area) / 10000;
    const result = !isNaN(currentAreaSize) && currentAreaSize ? areaSize / currentAreaSize : 0;
    return result;
  }
};

const getAverageScaleFactor = (measurements: ScaleMeasurement[]): number => {
  const scaleFactorMeasurements = measurements.reduce((prevScale, nextMeasurement) => {
    const nextScaleMeasurement = calculateScale(nextMeasurement);
    return prevScale + nextScaleMeasurement;
  }, 0);
  return scaleFactorMeasurements / measurements.length;
};

const calculateScaleFactorFromPageSpecs = (scaleRatio, paperFormat, sceneBackground) => {
  const currentPageSizes = PAGE_SIZES[paperFormat];
  const widthPxPerCm = sceneBackground.width / currentPageSizes.width;
  const heightPxPerCm = sceneBackground.height / currentPageSizes.height;

  // pixels per centimeters
  const pageFormat = (widthPxPerCm + heightPxPerCm) / 2;

  const scaleCm = scaleRatio / pageFormat;
  const scaleFactor = scaleCm ** 2;

  return scaleFactor;
};

const getPointsFromScaleToolLines = state => {
  const selectedLayer = getSelectedLayer(state.scene);
  const lines = Object.values(selectedLayer.lines);
  const points = lines
    .filter((line: any) => line.type === SeparatorsType.SCALE_TOOL)
    .reduce((accum: any[], line: any) => {
      const vertices = line.vertices.map(vertexID => {
        const vertex = selectedLayer.vertices[vertexID];
        return { x: vertex.x, y: vertex.y };
      });

      const orderedVertices = orderVertices(vertices);
      return [...accum, ...orderedVertices];
    }, []);
  return points;
};

export {
  isMeasurementOutlier,
  calculateScale,
  calculateScaleFactorFromPageSpecs,
  getAverageScaleFactor,
  needsRepeating,
  getPointsFromScaleToolLines,
};
