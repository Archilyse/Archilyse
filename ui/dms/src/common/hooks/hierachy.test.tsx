import { renderHook } from '@testing-library/react-hooks';
import { MOCK_AUTHENTICATION } from '../../../tests/utils';
import { useHierarchy } from '.';

jest.mock('react-router-dom', () => ({
  useLocation() {
    return {
      pathname: '/units',
      search: '?floor_id=12300',
      hash: '',
      key: 'c6t6fp',
    };
  },
  useParams() {
    return jest.fn();
  },
  useHistory() {
    return jest.fn();
  },
  Link() {
    return jest.fn();
  },
}));

it('Should display a hierarchy of entities from units for admin', async () => {
  MOCK_AUTHENTICATION();

  const { result, waitForNextUpdate } = renderHook(() => useHierarchy());
  await waitForNextUpdate();
  expect(result.current).toMatchSnapshot();
});
