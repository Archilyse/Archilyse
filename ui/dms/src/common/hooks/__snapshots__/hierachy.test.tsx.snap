// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should display a hierarchy of entities from units for admin 1`] = `
Array [
  Object {
    "href": "/clients",
    "text": "All Clients",
  },
  Object {
    "href": "/sites?client_id=63",
    "parent": [Function],
    "text": " A marvelous client",
  },
  Object {
    "href": "/buildings?site_id=1455",
    "parent": [Function],
    "text": " Site: Some office",
  },
  Object {
    "href": "/floors?building_id=2669",
    "parent": [Function],
    "text": " Building: Gotham, A",
  },
  Object {
    "href": "/units?floor_id=12300",
    "parent": [Function],
    "payload": Object {
      "building_id": 2669,
      "client_unit_mapping_link": null,
      "created": "2020-07-17T09:23:25.893567",
      "floor_number": 1,
      "gcs_de_dxf_link": null,
      "gcs_de_floorplan_link": null,
      "gcs_en_dxf_link": null,
      "gcs_en_floorplan_link": null,
      "gcs_heat_map_buildings_link": null,
      "gcs_heat_map_greenery_link": null,
      "gcs_heat_map_ground_link": null,
      "gcs_heat_map_isovist_link": null,
      "gcs_heat_map_mountains_link": null,
      "gcs_heat_map_railway_tracks_link": null,
      "gcs_heat_map_site_view_link": null,
      "gcs_heat_map_sky_view_link": null,
      "gcs_heat_map_streets_link": null,
      "gcs_heat_map_sun_december_afternoon_link": null,
      "gcs_heat_map_sun_december_midday_link": null,
      "gcs_heat_map_sun_december_morning_link": null,
      "gcs_heat_map_sun_june_afternoon_link": null,
      "gcs_heat_map_sun_june_midday_link": null,
      "gcs_heat_map_sun_june_morning_link": null,
      "gcs_heat_map_sun_march_afternoon_link": null,
      "gcs_heat_map_sun_march_midday_link": null,
      "gcs_heat_map_sun_march_morning_link": null,
      "gcs_heat_map_water_link": null,
      "georef_z": null,
      "id": 12300,
      "plan_id": 7514,
      "updated": null,
    },
    "text": "Floor 1",
  },
]
`;
